### ISSUE TITLE
parallelize select_mock_targets

### ISSUE THREAD

**Issue by forero** [desihub/desitarget#159]
Current implementation is parallel within a computing node using `multiprocessing`, but only handles a few square degrees.
A new parallel version should allow for the whole footprint creation using multiple nodes.

**Comment by forero** [comment]
@sbailey @moustakas I want to hear your opinions.

Right now I have three options to implement the parallelization. Before that, some numbers to keep in mind: for every 10 sqdeg we use ~10GB of RAM (~50GB as peak usage while reading the full mocks).

**1 Keep the code as it is now and use multiple input configuration files with different cuts in ra/dec for different jobs.**
Pros: Don't have to change the current code.
Cons:  Have to glue outputs for bricks that ended up in different ra/dec input regions. Multiple processes will try to access the mock files at the same time. For the bricks that were split into different input regions the target fluctuations won't be self-consistent.

**2 Split the mocks into bricks right after reading them. Use MPI to send different bricks to different processors, select targets on them and write the answer.**
Pros: Self consistent density fluctuations per brick.
Cons: Sending things by MPI might be slow. Have to actually change the code to manage all the MPI stuff.

**3 Split the mocks into bricks before reading them, i.e. create many brick files on disk and run them with different input configurations**
Pros: Finally, trivially parallel. Self-consistent target fluctuations on bricks. Not many changes in target selection.
Cons: Additional steps to prepare the data. New code to prepare the data.

**Comment by sbailey** [comment]
I think some version of (2) is best.  I chatted with @tskisner and speced out a proposal:
  * Calculate which MPI ranks are responsible for which bricks.  All ranks know which bricks are theirs.
  * Rank 0 reads the file(s) and broadcasts (`comm.Bcast`) in chunks of N targets
    * Ted says broadcasting all data to everyone is likely more efficient than having the master process
       try to do the filtering ahead of time and do targeted point-to-point send/receive.
  * Each rank receives the N targets into a buffer that it keeps reusing so that memory doesn't blow up
  * Each rank filters for the targets that are in its bricks and copies those.  This allows the filtering to be parallelized instead of having many ranks idle while one rank reads files and groups by brick before starting to send data to other ranks
  * After all ranks have received and filtered all data, they proceed with doing their brick calculations.

This is clearer in my head than it seems to be in that text so we may need to iterate. @tskisner offered to write a dummy example of the basic steps if we want to take him up on that offer.  I think the main pain will be that each mock is in a different format with a different set of columns such that it isn't a uniform thing that is being broadcast every time.

Let's try to keep the MPI/parallelism logic cleanly separated from the rest of the logic.

**Comment by forero** [comment]
OK. A dummy example from @tskisner would be indeed very helpful!

**Comment by moustakas** [comment]
I guess I was thinking of a much simpler version -- a variant of option 1 that doesn't have any of the cons.  Specifically, I think we just need to:
* Add an ra,dec `bounds` optional input to `mock.targets_truth` (which is currently optionally read from the input yaml file).
* Have a top-level wrapper script which basically uses `desispec.brick.brickname` to split the sky up into large "bricks" (e.g., 10 or 20 degrees on a side).  We'll want to check that the boundaries of these super-bricks coincide with the boundaries of the smaller bricks we define below.
* Generate N slurm scripts, one for each super-brick and call `mock.targets_truth` with the appropriate ra,dec bounds, making sure that `outbricksize` is an integer multiple of the size of each super-brick.
* Submit all the scripts with multiprocessing enabled to utilize all the available cores on a given node.

The only major con I can see is that the on-disk mocks will indeed be accessed by each node, but that's probably OK.

**Comment by sbailey** [comment]
I admit that the option Ted and I scoped out is somewhat heavyweight to implement, but a few challenges with @moustakas proposal:

The imaging surveys defined bricks don't tile into superbricks, so `desispec.brick.brickname` isn't useful for that.  If we redefine something that does tile, we shouldn't call it a "brick", and we should consider whether the effort to avoid healpix and/or HTM is more effort than just using one of those schemes.

With the current code, 10x10 deg^2 runs out of memory on an edison node.  3x3 deg^2 works, but that gives us ~2000 units of work to cover the DESI footprint, which would melt the file system (I'm guessing...) if they were all trying to independently read the multi-GB mock files simultaneously.  Either some level of reading and broadcasting is needed or significantly reducing the memory footprint so that much larger chunks could be done on a single node (which would independently be useful; it isn't clear to me why the current code uses as much memory as it does).

**Comment by sbailey** [comment]
Having just dissed on @moustakas suggestion, I will now proceed by admitting that the basic concept of moving RA,dec bounds into `mock.targets_truth` and calling it a bunch of times for different patches of sky was basically what I was originally thinking about when we started discussing this.  I'll add one idea to that that could help make this more viable for reading files:

For each flavor of mock, build an index file with:
* HDU 1 EXTNAME='FILEMAP'
  * header keyword BASEDIR = base directory to prepend to filenames in the data table
  * data table with columns FILENUM (int16) and FILENAME (str*) with one row per file in these mocks
* HDU 2 EXTNAME='TARGETINDEX'
  * data table with columns FILENUM (int16), RA (float64), DEC (float64), and maybe MOCKID (int64)

The mock builder could start by reading the index, filtering on RA,DEC, and then using FILENUM -> BASEDIR+FILENAME to look up just the subset of files that are actually needed for this piece of sky.  This could also be useful for the target -> mock provenance described in #88.

This index file doesn't apply so well to the Gaussian mocks that are already in a single file per target class with minimal columns, but it could be useful for BGS, MWS, and LYA where you don't really want to have to read all the files if you only need a small subset of them.

I'm still not sure this is viable for N>>1 parallel processes reading the same files, but I thought I'd throw it out there.

**Comment by sbailey** [comment]
One other tidbit that might be useful (sorry for all the emails...)

fitsio has the ability to efficiently filter on rows of a table to read in only a subset, e.g.
```python
with fitsio.FITS('blat.fits') as fx:
    w = fx[1].where('200 <= RA && RA < 210 && 0 <= DEC && DEC < 10')
    data = fx[1][w]
```

**Comment by moustakas** [comment]
That's a nice trick, thanks @sbailey.  

Another option would be to (approximately) parallelize over spectral type, which would require of order 10ish yaml files (which could even be created on-the-fly).  For example, could we do the ELGs all-sky with 24 cores in a reasonable amount of time (after investigating all the bottlenecks and possible speed-ups), then the MWS, then the LRGs, etc.  ?  I'll try to do some tests unless someone beats me to it.

**Comment by forero** [comment]
The index file is definitely useful for 'MWS' and 'LYA' all the other mocks (gaussian + BGS) come as a monolithic file. 

You both agree on the idea of filtering by RA,DEC. Does that mean that we drop the brick splitting? If not, we have to deal with concurrent processes writing to the same brick file.

**Comment by sbailey** [comment]
Agreed; let's not try to have different processes write to the same brick.  We either split by bricks, or split by something else that isn't a "brick" (e.g. larger RA,dec bounds), but coordinating half a brick in one process and half a brick in another while trying to write them to the same file would be a huge pain.

**Comment by forero** [comment]
just a heads-up after a quick chat with Stephen,  I will work on saturday on a toy implementation of [Stephen's suggestion](https://github.com/desihub/desitarget/issues/159#issuecomment-286604472). I will let you know about my progress and whether I need @tskisner expertise to move faster towards a solution.

**Comment by rainwoodman** [comment]
If the expected density can be looked up trivially from ra dec, then you
ran throw a random for each objtype and decide to keep or to drop. This way
there is no need to split the full mock to bricks.

Yu

**Comment by forero** [comment]
Right, Yu. That's part of the new strategy. Thanks.

**Comment by tskisner** [comment]
Maybe it is not relevant any more for this discussion, but here is the small example I was talking about with Stephen.  This code is not tested at all and has placeholder function calls, but you get the idea:

[https://gist.github.com/tskisner/ebf41a77264dc721c9c098d83eb11f37](url)

**Comment by moustakas** [comment]
Implemented by @sbailey in #188.
